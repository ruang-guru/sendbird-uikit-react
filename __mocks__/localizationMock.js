export const STRING_SET = {
  BUTTON__READ_MORE: 'Baca selengkapnya',
  CHANNEL__MESSAGE_LIST__NOTIFICATION__ON: 'pada',
  CHANNEL__MESSAGE_LIST__NOTIFICATION__NEW_MESSAGE: 'pesan baru',
  LABEL__OPERATOR: 'Guru',
  LABEL__DATE_TODAY: 'Hari ini',
  LABEL__DATE_YESTERDAY: 'Kemarin',
  LABEL__DAY_SUNDAY: 'Minggu',
  LABEL__DAY_MONDAY: 'Senin',
  LABEL__DAY_TUESDAY: 'Selasa',
  LABEL__DAY_WEDNESDAY: 'Rabu',
  LABEL__DAY_THURSDAY: 'Kamis',
  LABEL__DAY_FRIDAY: 'Jumat',
  LABEL__DAY_SATURDAY: 'Sabtu',
};

export default STRING_SET;
