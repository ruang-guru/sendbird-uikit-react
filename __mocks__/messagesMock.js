export const USER_ID_A = 'user_id_a';
export const BASIC_MESSAGE = {
  messageId: 1479690277,
  messageType: 'user',
  channelUrl:
    'sendbird_group_channel_337761017_9f2bf745b4e1d55a20c4b33a8f3a1cb6daecf47d',
  data: '',
  customType: '',
  silent: false,
  createdAt: 1632383106322,
  updatedAt: 0,
  channelType: 'group',
  metaArrays: [],
  reactions: [],
  mentionType: 'users',
  mentionedUsers: [],
  sendingStatus: 'succeeded',
  parentMessageId: 0,
  parentMessageText: null,
  threadInfo: {
    replyCount: 0,
    mostRepliedUsers: [],
    lastRepliedAt: 0,
    updatedAt: 0,
  },
  ogMetaData: null,
  isOperatorMessage: false,
  appleCriticalAlertOptions: null,
  message: 'Welcome to the classroom!',
  sender: {
    nickname: 'Kukuh Aji Sulistyo Bangun Jaya Abadi Mangkubumi',
    plainProfileUrl:
      'https://avatars3.githubusercontent.com/u/46333979?s=460&v=4',
    profileUrl: 'https://avatars3.githubusercontent.com/u/46333979?s=460&v=4',
    userId: 'random-user-id-xxx',
    connectionStatus: 'nonavailable',
    lastSeenAt: 0,
    metaData: {},
    isActive: true,
    friendDiscoveryKey: null,
    friendName: null,
    _preferredLanguages: null,
    requireAuth: false,
    role: 'none',
    isBlockedByMe: false,
  },
  reqId: '1632382917015',
  translations: {},
  requestState: 'succeeded',
  requestedMentionUserIds: [],
  errorCode: 0,
  messageSurvivalSeconds: -1,
  plugins: [],
  poll: null,
};

export const BASIC_MESSAGE_A_1 = {
  ...BASIC_MESSAGE,
  sender: { ...BASIC_MESSAGE.sender, userId: USER_ID_A },
};
export const BASIC_MESSAGE_A_2 = {
  ...BASIC_MESSAGE,
  message:
    'Can you think of other strategies for ensuring a welcoming first day of school? Do you have ideas that you have used successfully over the years? Tell us about them n the comments field below.',
  sender: { ...BASIC_MESSAGE.sender, userId: USER_ID_A },
};
export const BASIC_MESSAGE_A_3 = {
  ...BASIC_MESSAGE,
  message: 'Of course, outside reading log',
  sender: { ...BASIC_MESSAGE.sender, userId: USER_ID_A },
};

export const ASSIGNMENT_MESSAGE_A_1 = {
  ...BASIC_MESSAGE,
  customType: 'assignment',
  data: "{\"serial\":\"ASN-F8ES2S32KFTAR5S6\",\"title\":\"Pilihan Ganda Template with deadline\",\"description\":\"Pilihan Ganda Template\",\"dueAt\":\"2022-01-01T13:20:00Z\",\"creatorSerial\":\"JILLYU9G79PUXOHK\",\"cta\":\"ruangguru://ruangkelas?page=assignment_detail\\u0026workspace_serial=LMS-WS-YPJJS4QTPQU89LR1\\u0026classroom_serial=LMS-CR-ARHSS4XK3RP1R0HQ\\u0026assignment_serial=ASN-F8ES2S32KFTAR5S6\",\"ctaWeb\":\"https://kelas.sirogu.com/workspace/LMS-WS-YPJJS4QTPQU89LR1/classroom/LMS-CR-ARHSS4XK3RP1R0HQ/assignment/detail/ASN-F8ES2S32KFTAR5S6\"}",
};

export const ASSIGNMENT_MESSAGE_A_2 = {
  ...ASSIGNMENT_MESSAGE_A_1,
  data: "{\"serial\":\"ASN-IHB4XI1KY1HWRWPO\",\"title\":\"Pilihan Ganda Template\",\"description\":\"Pilihan Ganda Template\",\"dueAt\":\"\",\"creatorSerial\":\"JILLYU9G79PUXOHK\",\"cta\":\"ruangguru://ruangkelas?page=assignment_detail\\u0026workspace_serial=LMS-WS-YPJJS4QTPQU89LR1\\u0026classroom_serial=LMS-CR-ARHSS4XK3RP1R0HQ\\u0026assignment_serial=ASN-IHB4XI1KY1HWRWPO\",\"ctaWeb\":\"https://kelas.sirogu.com/workspace/LMS-WS-YPJJS4QTPQU89LR1/classroom/LMS-CR-ARHSS4XK3RP1R0HQ/assignment/detail/ASN-IHB4XI1KY1HWRWPO\"}",
  sender: { ...BASIC_MESSAGE.sender, userId: USER_ID_A },
};

export const MATERIAL_MESSAGE_A_1 = {
  ...BASIC_MESSAGE,
  customType: 'material',
  data: "{\"serial\":\"SUBTOPI-2N28ZO17\",\"title\":\"Materi Google\",\"creatorSerial\":\"JILLYU9G79PUXOHK\",\"description\":\"\",\"cta\":\"ruangguru://ruangkelas?page=material_detail\\u0026workspace_serial=LMS-WS-YPJJS4QTPQU89LR1\\u0026classroom_serial=LMS-CR-ARHSS4XK3RP1R0HQ\\u0026material_serial=SUBTOPI-2N28ZO17\",\"ctaWeb\":\"https://kelas.sirogu.com/workspace/LMS-WS-YPJJS4QTPQU89LR1/classroom/LMS-CR-ARHSS4XK3RP1R0HQ/material/detail/SUBTOPI-2N28ZO17\"}"
};

export const MATERIAL_MESSAGE_A_2 = {
  ...MATERIAL_MESSAGE_A_1,
  data: "{\"serial\":\"SUBTOPI-2N28ZO17\",\"title\":\"Understand how users behave on your site, what they need,  Understand how users behave on your site, and how they feel, fast \",\"creatorSerial\":\"JILLYU9G79PUXOHK\",\"description\":\"\",\"cta\":\"ruangguru://ruangkelas?page=material_detail\\u0026workspace_serial=LMS-WS-YPJJS4QTPQU89LR1\\u0026classroom_serial=LMS-CR-ARHSS4XK3RP1R0HQ\\u0026material_serial=SUBTOPI-2N28ZO17\",\"ctaWeb\":\"https://kelas.sirogu.com/workspace/LMS-WS-YPJJS4QTPQU89LR1/classroom/LMS-CR-ARHSS4XK3RP1R0HQ/material/detail/SUBTOPI-2N28ZO17\"}",
  sender: { ...BASIC_MESSAGE.sender, userId: USER_ID_A },
};
export const LONG_MESSAGE = {
  ...BASIC_MESSAGE,
  message:
    'Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going through the cites of the word in classical literature, discovered the undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of "de Finibus Bonorum et Malorum" (The Extremes of Good and Evil) by Cicero, written in 45 BC. This book is a treatise on the theory of ethics, very popular during the Renaissance. The first line of Lorem Ipsum, "Lorem ipsum dolor sit amet..", There are many variations of passages of Lorem Ipsum available, but the majority have suffered alteration in some form, by injected humour, or randomised words which don\'t look even slightly believable. If you are going to use a passage of Lorem Ipsum, you need to be sure there isn\'t anything embarrassing hidden in the middle of text. All the Lorem Ipsum generators on the Internet tend to repeat predefined chunks as necessary, making this the first true generator on the Internet.',
};
export const LONG_MESSAGE_A_1 = {
  ...LONG_MESSAGE,
  sender: { ...LONG_MESSAGE.sender, userId: USER_ID_A },
};

export const OPERATOR_MESSAGE = {
  ...BASIC_MESSAGE,
  isOperatorMessage: true,
};
