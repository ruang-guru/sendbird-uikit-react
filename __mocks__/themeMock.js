export default {
  '--sendbird-light-primary-100': '#FFFAE6',

  '--sendbird-light-background-100': '#E6EBF1',

  '--sendbird-light-onlight-01': '#434856',
  '--sendbird-light-secondary-300': '#2EB5C0',
};
