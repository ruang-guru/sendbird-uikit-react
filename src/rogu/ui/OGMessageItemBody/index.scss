@import '../../../styles/variables';

.rogu-og-message-item-body {
  position: relative;
  display: inline-block;
  width: 100%;
  min-width: 200px;
  max-width: 400px;

  .rogu-og-message-item-body__og-wrapper {
    padding: 8px;
    display: flex;
    border-radius: 4px;
    margin-bottom: 8px;
  }

  .rogu-og-message-item-body__og-container {
    display: flex;
    position: relative;
    width: auto;
  }

  .rogu-og-message-item-body__og-thumbnail {
    position: relative;
    width: 60px;
    height: 60px;

    .rogu-og-message-item-body__og-thumbnail__image {
      width: 100%;
    }
  }

  .rogu-og-message-item-body__description {
    padding: 4px 8px;
    flex: 1;
    width: 100px;

    .rogu-og-message-item-body__description__title,
    .rogu-og-message-item-body__description__description {
      margin-bottom: 4px;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
      display: block;
    }

    .rogu-og-message-item-body__description__url {
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
      display: block;
      margin: 0;
    }
  }

  &.rogu-og-message-item-body--preview {
    max-width: 100%;

    .rogu-og-message-item-body__og-wrapper {
      align-items: center;
      justify-content: space-between;
    }

    .rogu-og-message-item-body__og-container {
      width: 90%;
    }
  }

  &.rogu-og-message-reactions .rogu-og-message-item-body__description {
    border-radius: 0;
  }

  .rogu-og-message-item-body__og-thumbnail:hover,
  .rogu-og-message-item-body__description:hover {
    cursor: pointer;
  }

  &.rogu-og-message--outgoing {
    .rogu-og-message-item-body__og-wrapper {
      @include themed() {
        background-color: t(primary-2);
      }
    }
  }

  &.rogu-og-message--incoming {
    .rogu-og-message-item-body__og-wrapper {
      @include themed() {
        background-color: t(bg-1);
      }
    }
  }

  &.mouse-hover,
  &:hover {
    .rogu-og-message--outgoing {
      .rogu-og-message-item-body__og-wrapper {
        @include themed() {
          background-color: t(primary-2);
        }
      }
    }

    .rogu-og-message--incoming {
      .rogu-og-message-item-body__og-wrapper {
        @include themed() {
          background-color: t(bg-1);
        }
      }
    }
  }
}
