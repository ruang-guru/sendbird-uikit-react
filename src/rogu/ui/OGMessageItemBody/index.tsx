import React, { ReactElement, useContext } from 'react';
import { UserMessage } from 'sendbird';
import './index.scss';

import LinkLabel from '../../../ui/LinkLabel';
import ImageRenderer from '../../../ui/ImageRenderer';

import {
  getClassName,
  isEditedMessage,
  isUrl,
} from '../../../utils';
import uuidv4 from '../../../utils/uuid';
import { LocalizationContext } from '../../../lib/LocalizationContext';

import Label, { LabelTypography, LabelColors } from '../Label';
import Icon, { IconTypes, IconColors } from '../Icon';

import IconButton from '../IconButton';

interface Props {
  className?: string | Array<string>;
  message: UserMessage;
  isByMe?: boolean;
  mouseHover?: boolean;
  isOnPreview?: boolean;
  onClosePreview?: () => void;
}

export default function OGMessageItemBody({
  className,
  message,
  isByMe = false,
  mouseHover = false,
  isOnPreview = false,
  onClosePreview,
}: Props): ReactElement {
  const { stringSet } = useContext(LocalizationContext);
  const openOGUrl = (): void => {
    if (message?.ogMetaData?.url) window.open(message?.ogMetaData?.url);
  };

  return (
    <div className={getClassName([
      className,
      'rogu-og-message-item-body',
      isByMe ? 'rogu-og-message--outgoing' : 'rogu-og-message--incoming',
      isOnPreview ? 'rogu-og-message-item-body--preview' : '',
      mouseHover ? 'mouse-hover' : '',
      message?.reactions?.length > 0 ? 'rogu-og-message-reactions' : '',
    ])}>

      <div className="rogu-og-message-item-body__og-wrapper" >
        <div className="rogu-og-message-item-body__og-container" onClick={openOGUrl}>
          <div
            className="rogu-og-message-item-body__og-thumbnail"
          >
            <ImageRenderer
              className="rogu-og-message-item-body__og-thumbnail__image"
              url={message?.ogMetaData?.defaultImage?.url || ''}
              alt={message?.ogMetaData?.defaultImage?.alt}
              width="60px"
              height="60px"
              defaultComponent={(
                <div className="rogu-og-message-item-body__og-thumbnail__place-holder">
                  <Icon
                    className="rogu-og-message-item-body__og-thumbnail__place-holder__icon"
                    type={IconTypes.THUMBNAIL_NONE}
                    width="60px"
                    height="60px"
                  />
                </div>
              )}
            />
          </div>
          <div className="rogu-og-message-item-body__description">
            {message?.ogMetaData?.title && (
              <Label
                className="rogu-og-message-item-body__description__title"
                type={LabelTypography.SUBTITLE_2}
                color={LabelColors.ONBACKGROUND_1}
              >
                {message.ogMetaData.title}
              </Label>
            )}
            {message?.ogMetaData?.description && (
              <Label
                className="rogu-og-message-item-body__description__description"
                type={LabelTypography.BODY_2}
                color={LabelColors.ONBACKGROUND_1}
              >
                {message.ogMetaData.description}
              </Label>
            )}
            {message?.ogMetaData?.url && (
              <Label
                className="rogu-og-message-item-body__description__url"
                type={LabelTypography.CAPTION_3}
                color={LabelColors.ONBACKGROUND_2}
              >
                {message.ogMetaData.url}
              </Label>
            )}
          </div>
        </div>
        {
          isOnPreview && <IconButton
            className="sendbird-chat-header__right__search"
            width="32px"
            height="32px"
            onClick={onClosePreview}
          >
            <Icon
              type={IconTypes.CLOSE}
              fillColor={IconColors.ON_BACKGROUND_1}
              width="24px"
              height="24px"
            />
          </IconButton>
        }
      </div>

      <div className="rogu-og-message-item-body__text-bubble">
        {message?.message.split(' ').map((word: string) => (
          isUrl(word)
            ? (
              <LinkLabel
                className="rogu-og-message-item-body__text-bubble__message"
                key={uuidv4()}
                src={word}
                type={LabelTypography.BODY_1}
                color={isByMe ? LabelColors.ONBACKGROUND_1 : LabelColors.SECONDARY_3}
              >
                {word}
              </LinkLabel>
            )
            : (
              <Label
                className="rogu-og-message-item-body__text-bubble__message"
                key={uuidv4()}
                type={LabelTypography.BODY_1}
                color={LabelColors.ONBACKGROUND_1}
              >
                {word + ' '}
              </Label>
            )
        )
        )}
        {isEditedMessage(message) && (
          <Label
            className="rogu-og-message-item-body__text-bubble__message"
            type={LabelTypography.BODY_1}
            color={isByMe ? LabelColors.ONCONTENT_2 : LabelColors.ONBACKGROUND_2}
          >
            {` ${stringSet.MESSAGE_EDITED} `}
          </Label>
        )}
      </div>
      <div className="rogu-og-message-item-body__cover" />
    </div>
  );
}
