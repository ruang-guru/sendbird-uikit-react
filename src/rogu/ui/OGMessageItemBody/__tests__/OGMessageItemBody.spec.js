import React from 'react';
import renderer from 'react-test-renderer';

import OGMessageItemBody from '../index';

import { OG_MESSAGE } from '../../../../../__mocks__/messagesMock';

describe('OGMessageItemBody', () => {
  it('should do a snapshot test of the OGMessageItemBody DOM', function () {
    const component = renderer.create(
      <OGMessageItemBody message={OG_MESSAGE} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
