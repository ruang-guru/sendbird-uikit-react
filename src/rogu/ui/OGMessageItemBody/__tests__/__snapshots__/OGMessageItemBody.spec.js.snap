// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OGMessageItemBody should do a snapshot test of the OGMessageItemBody DOM 1`] = `
<div
  className=" rogu-og-message-item-body rogu-og-message-incoming  "
>
  <div
    className="rogu-og-message-item-body__og-container"
    onClick={[Function]}
  >
    <div
      className="rogu-og-message-item-body__og-thumbnail"
    >
      <div
        className="rogu-og-message-item-body__og-thumbnail__image sendbird-image-renderer"
        style={
          Object {
            "height": "60px",
            "maxWidth": "400px",
            "minWidth": "60px",
            "width": "100%",
          }
        }
      >
        <div
          className="sendbird-image-renderer__image"
          style={
            Object {
              "backgroundImage": "url()",
              "backgroundPosition": "center",
              "backgroundRepeat": "no-repeat",
              "backgroundSize": "cover",
              "borderRadius": null,
              "height": "60px",
              "maxWidth": "400px",
              "minWidth": "60px",
              "position": "absolute",
              "width": "100%",
            }
          }
        />
        <img
          alt=""
          className="sendbird-image-renderer__hidden-image-loader"
          onError={[Function]}
          onLoad={[Function]}
          src=""
        />
      </div>
    </div>
    <div
      className="rogu-og-message-item-body__description"
    />
  </div>
  <div
    className="rogu-og-message-item-body__text-bubble"
  />
  <div
    className="rogu-og-message-item-body__cover"
  />
</div>
`;
